src/memo/memo.ml

```diff
@@ -215,20 +315,56 @@ module Cycle_error = struct
   let stack t = t.stack
 end
 
-module type S = Memo_intf.S with type stack_frame := Stack_frame.t
-
 let global_dep_dag = Dag.create ()
 
-(* fiber context variable keys *)
-let call_stack_key = Fiber.Var.create ()
-let get_call_stack () = Fiber.Var.get call_stack_key |> Option.value ~default:[]
+(* call stack consists of two components: asynchronous call stack managed with a fiber
+   context variable and synchronous call stack on top of that managed with an explicit ref *)
+module Call_stack = struct
+
+  let synchronous_call_stack = ref []
+
+  let list_first = function
```

Comment:
Can we move this to `List.hd_opt` (or w/e name you like)

Reply:
`hd_opt` is fine. Pushed.
I think I like simple `hd` better though. If you agree, I'll rename it.

Reply:
`List.hd` is good, but I think you will need to refactor a lot of code to use the old `List.hd` (which would presumably be `List.hd_exn`)

Reply:
I'm not doing it for now then.